plugin/pkg/admission/storage/volumeattributesclass/setdefault/admission.go

```diff
@@ -0,0 +1,120 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package setdefault
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"k8s.io/kubernetes/pkg/volume/util"
+
+	"k8s.io/klog/v2"
+
+	"k8s.io/apiserver/pkg/admission"
+	genericadmissioninitializer "k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/client-go/informers"
+	storagev1alpha1listers "k8s.io/client-go/listers/storage/v1alpha1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+)
+
+const (
+	// PluginName is the name of this admission controller plugin
+	PluginName = "DefaultVolumeAttributesClass"
+)
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		plugin := newPlugin()
+		return plugin, nil
+	})
+}
+
+// claimDefaulterPlugin holds state for and implements the admission plugin.
+type claimDefaulterPlugin struct {
+	*admission.Handler
+
+	lister storagev1alpha1listers.VolumeAttributesClassLister
+}
+
+var _ admission.Interface = &claimDefaulterPlugin{}
+var _ admission.MutationInterface = &claimDefaulterPlugin{}
+var _ = genericadmissioninitializer.WantsExternalKubeInformerFactory(&claimDefaulterPlugin{})
+
+// newPlugin creates a new admission plugin.
+func newPlugin() *claimDefaulterPlugin {
+	return &claimDefaulterPlugin{
+		Handler: admission.NewHandler(admission.Create),
+	}
+}
+
+func (a *claimDefaulterPlugin) SetExternalKubeInformerFactory(f informers.SharedInformerFactory) {
+	informer := f.Storage().V1alpha1().VolumeAttributesClasses()
+	a.lister = informer.Lister()
+	a.SetReadyFunc(informer.Informer().HasSynced)
+}
+
+// ValidateInitialization ensures lister is set.
+func (a *claimDefaulterPlugin) ValidateInitialization() error {
+	if a.lister == nil {
+		return fmt.Errorf("missing lister")
+	}
+	return nil
+}
+
+// Admit sets the default value of a PersistentVolumeClaim's volume attributes class, in case the user did
+// not provide a value.
+//
+// 1.  Find available VolumeAttributesClass.
+// 2.  Figure which is the default
+// 3.  Write to the PVClaim
+func (a *claimDefaulterPlugin) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != api.Resource("persistentvolumeclaims") {
+		return nil
+	}
+
+	if len(attr.GetSubresource()) != 0 {
+		return nil
+	}
+
+	pvc, ok := attr.GetObject().(*api.PersistentVolumeClaim)
+	// if we can't convert then we don't handle this object so just return
+	if !ok {
+		return nil
+	}
+
+	if pvc.Spec.VolumeAttributesClassName != nil {
+		// The user asked for a class.
+		return nil
+	}
+
+	klog.V(4).Infof("no volume attributes class for claim %s (generate: %s)", pvc.Name, pvc.GenerateName)
+
+	def, err := util.GetDefaultVolumeAttributesClass(a.lister)
```

Comment:
```text
A default VolumeAttributesClass can be specified for the Kubernetes cluster. This default VolumeAttributesClass is then used to dynamically provision storage for PersistentVolumeClaims that do not require any specific VolumeAttributesClass. A cluster admin can use annotation to manage default VolumeAttributesClass. The default VolumeAttributesClass has an annotation volumeattributesclass.kubernetes.io/is-default-class set to true. Any other value or absence of the annotation is interpreted as false.
```

Above content is from [kubernetes/enhancements#3751](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3751-volume-attributes-class#default-volumeattributesclass).

I've different understanding with it. I think one default attributes class per csi-driver, not per cluster. Because one default attributes class per cluster is not reasonable. For example, if I have two csi-drivers, one is for GCP, another is for NetApp and two VACs, one (gcp-vac) is for GCP , another (netapp-vac) is for NetApp.

I want to use gcp-vac as default attributes class for GCP csi-driver, and use netapp-vac as default attributes class for NetApp csi-driver. But if one default attributes class per cluster, I can't do it because the attributes of volumes between two SPs are different. 

On the other hand, when default SC and default VAC both exists in a cluster, if users create a PVC without SC and VAC specified, it's possible that the PVC will stuck in pending state forever. Because the default SC and default VAC may not be compatible if the csi-driver doesn't support both of them.

So I think one default attributes class per csi-driver is reasonable.

@sunnylovestiramisu @xing-yang what do you think?

Reply:
I think one default VAC per csi driver is preferable.

Reply:
@xing-yang @sunnylovestiramisu 

It's done in https://github.com/kubernetes/kubernetes/pull/118863/commits/43318dd3a26620d94dc8565c612590792d46ec4f